src/amm.rs

```diff
@@ -73,84 +110,69 @@ pub fn calculate_spot_price(sell_reserve: Balance, buy_reserve: Balance, amount:
 /// - `sell_amount` - amount
 ///
-/// Returns None in case of error
-pub fn calculate_sell_price(sell_reserve: Balance, buy_reserve: Balance, sell_amount: Balance) -> Option<Balance> {
+/// Returns MathError in case of error
+pub fn calculate_sell_price(sell_reserve: Balance, buy_reserve: Balance, sell_amount: Balance) -> Result<Balance, MathError> {
     let (sell_amount_hp, buy_reserve_hp, sell_reserve_hp) = to_u256!(sell_amount, buy_reserve, sell_reserve);
+    let divisor = add!(sell_reserve_hp, sell_amount_hp);
+    ensure!(!divisor.is_zero(), DivisorIsZero);
 
-    let numerator = buy_reserve_hp.checked_mul(sell_amount_hp)?;
-    let denominator = sell_reserve_hp.checked_add(sell_amount_hp)?;
+    let dividend = multiply!(buy_reserve_hp, sell_amount_hp);
+    let sale_price_hp = divide!(dividend, divisor);
 
-    let sale_price_hp = numerator.checked_div(denominator)?;
-
-    match to_u128!(sale_price_hp) {
-        Some(sale_price) => round_up!(sale_price),
-        None => None,
-    }
+    let result = to_u128!(sale_price_hp).ok_or(ResultOverflow).ok();
+    round_up!(result.unwrap()).ok_or(ResultOverflow)
 }
 
 /// Calculating buying price given reserve of selling asset and reserve of buying asset.
-/// Formula : SELL_RESERVE * AMOUNT / (BUY_RESERVE - AMOUNT )
+/// Formula : SELL_RESERVE * AMOUNT / (BUY_RESERVE - AMOUNT)
 ///
 /// - `sell_reserve` - reserve amount of selling asset
 /// - `buy_reserve` - reserve amount of buying asset
 /// - `amount` - buy amount
 ///
 /// Returns None in case of error
-pub fn calculate_buy_price(sell_reserve: Balance, buy_reserve: Balance, amount: Balance) -> Option<Balance> {
-    ensure!(amount <= buy_reserve);
+pub fn calculate_buy_price(sell_reserve: Balance, buy_reserve: Balance, amount: Balance) -> Result<Balance, MathError> {
+    ensure!(amount <= buy_reserve, InsufficientBuyReserve);
 
     let (amount_hp, buy_reserve_hp, sell_reserve_hp) = to_u256!(amount, buy_reserve, sell_reserve);
+    let divisor = substract!(buy_reserve_hp, amount_hp);
+    ensure!(!divisor.is_zero(), DivisorIsZero);
 
-    let numerator = sell_reserve_hp.checked_mul(amount_hp)?;
-    let denominator = buy_reserve_hp.checked_sub(amount_hp)?;
-    let buy_price_hp = numerator.checked_div(denominator)?;
+    let dividend = multiply!(sell_reserve_hp, amount_hp);
+    let buy_price_hp = divide!(dividend, divisor);
 
-    match to_u128!(buy_price_hp) {
-        Some(buy_price) => round_up!(buy_price),
-        None => None,
-    }
+    let result = to_u128!(buy_price_hp).ok_or(ResultOverflow).ok();
+    round_up!(result.unwrap()).ok_or(ResultOverflow)
 }
 
-pub fn calculate_liquidity_in(asset_a_reserve: Balance, asset_b_reserve: Balance, amount: Balance) -> Option<Balance> {
-    ensure!(asset_a_reserve != 0);
+pub fn calculate_liquidity_in(asset_a_reserve: Balance, asset_b_reserve: Balance, amount: Balance) -> Result<Balance, MathError> {
+    ensure!(asset_a_reserve != 0, DivisorIsZero);
```

Comment:
If you agree, I would probably introduce new Errors like maybe "AssetAReserveIsZero" and "LiquidityIsZero".
While "DivisorIsZero" is true here, we would again force the client to think about details of this function.
Maybe we can gain some Expressiveness here.

Reply:
No, like we said - return 0. No error needed